### PR TITLE
Feat(lang): add addons gitlab support

### PR DESCRIPTION
**Pull request by lgj101**
### Description of your changes

Add kubevela addons gitlab support

Fixes #

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [x] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested
Create addons in the private gitlab repository，
Example：https://example.gitlab.com/xxx/xxx/kube-addon

Request kubevela apiserver locally, create gitlab repository addons, and obtain addons in gitlab repository
Example:
Post Request : http://localhost:8000/api/v1/addon_registries 
Params:
{
    "name":"test1",
    "gitlab":{
        "url":"https://example.gitlab.com/beijing/kube/kube-addon",
        "token":"xxxxxxxxx",
        "path":"local-addons",
        "owner":"/beijing/kube"
    }
}

Response:
{
    "name":"test1",
    "gitlab":{
        "url":"https://example.gitlab.com/beijing/kube/kube-addon",
        "token":"xxxxxxxxx",
        "path":"local-addons",
        "owner":"/beijing/kube"
    }
}

Get Request : http://localhost:8000/api/v1/addons?registry=test1
Response:
{
    "addons":[
        Object{...},
        Object{...}
    ]
}

### CODE REVIEW

**Comment by msftgits** [comment]
[CLA assistant check](https://cla.opensource.microsoft.com/oam-dev/kubevela?pullRequest=3543) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/oam-dev/kubevela?pullRequest=3543) before we can accept your contribution.<br/><br/>:x: lgj101 [sign now](https://cla.opensource.microsoft.com/oam-dev/kubevela?pullRequest=3543)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/oam-dev/kubevela?pullRequest=3543) it.</sub>

File: pkg/addon/source.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ package addon
 
 import (
 	"fmt"
+	"github.com/xanzy/go-gitlab"
```
Comment (chivalryq) [inline pkg/addon/source.go:21]: Please move to third party imports block

File: pkg/utils/parse.go [hunk 2]
```diff
@@ -198,3 +212,24 @@ func ByteCountIEC(b int64) string {
 	return fmt.Sprintf("%.1f %ciB",
 		float64(b)/float64(div), "KMGTPE"[exp])
 }
+
+// ParseGitlab will parse gitlab config from address
+func ParseGitlab(addr, owner string) (string, *Content, error) {
+	// We support two valid format:
```
Comment (chivalryq) [inline pkg/utils/parse.go:218]: Maybe One format?

File: pkg/addon/source.go [hunk 3]
```diff
@@ -35,6 +36,10 @@ const (
 	DirType = "dir"
 	// FileType means a file
 	FileType = "file"
+	// BlobType means a blob
+	BlobType = "blob"
+	// TreeType means a blob
```
Comment (chivalryq) [inline pkg/addon/source.go:41]: typo

File: pkg/addon/source.go (hunk 1 above)
Comment (lgj101) [reply pkg/addon/source.go:21]: 第三方的 放在哪里呢  有没有示例参考一下

File: pkg/utils/parse.go (hunk 2 above)
Comment (lgj101) [reply pkg/utils/parse.go:218]: gitlab仓库地址的host和owner层级 是不确定的，不能采用同样github和gitee的方式去解析url，你这里说的没太理解，能详细说说吗

File: pkg/addon/reader_gitlab.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+var _ AsyncReader = &gitlabReader{}
+
+// gitlabReader helps get addon's file by git
+type gitlabReader struct {
+	h *gitlabHelper
+}
+
+// gitlabHelper helps get addon's file by git
+type gitlabHelper struct {
```
Comment (wangyikewxgm) [inline pkg/addon/reader_gitlab.go:35]: Please add test for this implementation like other readers.

Comment (chivalryq) [reply pkg/addon/reader_gitlab.go:35]: Just a hint, you need mock a server which will return response as GitlLab does.

Comment (lgj101) [reply pkg/addon/reader_gitlab.go:35]: OK，Let me see how to write it

Comment (lgj101) [reply pkg/addon/reader_gitlab.go:35]: 1. 我的这个gitlabhepler实现是借助第三方包"github.com/xanzy/go-gitlab"，client是第三方包的，所以它的reader_gitlab_test文件不支持本地创建httpserver，然后监听扫描本地testdata文件，它会直接访问外部
2. gitlab的api访问和github以及gitee的访问方式是有区别的，需要先拿到仓库的project id，然后根据project id去找到仓库进行扫描。
我创建了自己的gitlab公开插件仓库，但是我觉得用我的肯定是不合适的。如果可以的话，你们是不是可以提供一个公用的gitlab插件仓库，以供reader_gitlab_test文件去扫描。
如果这样不可以的话，是不能借助第三方包去实现才能够通过测试吗

Comment (lgj101) [reply pkg/addon/reader_gitlab.go:35]: > Just a hint, you need mock a server which will return response as GitlLab does.

这个是必须的吗

Comment (chivalryq) [reply pkg/addon/reader_gitlab.go:35]: @lgj101  Hi, using a package from the third party is fine. Talking to test, maybe you could build the `Client` yourself? Ref to how github reader is tested.
Build a Server and pass the client to gitlabHelper. Then this client can send http request to a mock server.
Ref more detail to `setup()` in `pkg/addon/reader_github_test.go`. Feel free to ack if you still have problems

Comment (chivalryq) [reply pkg/addon/reader_gitlab.go:35]: We need to keep the function tested in `/pkg`. Add tests so you can more confirmed about how GitLab API works so maybe you can find some hidden bugs

Comment (lgj101) [reply pkg/addon/reader_gitlab.go:35]: > We need to keep the function tested in `/pkg`. Add tests so you can more confirmed about how GitLab API works. Maybe this helps you to find some hidden bugs

好的 我尝试一下

File: pkg/addon/reader_gitlab_test.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	"net/http/httptest"
+	"path"
+	"strings"
+	"testing"
+
+	"github.com/xanzy/go-gitlab"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/oam-dev/kubevela/pkg/utils"
```
Comment (chivalryq) [inline pkg/addon/reader_gitlab_test.go:32]: Please sorting all imports by
1. built-ins
2. from oam-dev/kubevela
3. from third party

File: pkg/addon/reader_gitlab_test.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	// mux is the HTTP request multiplexer used with the test server.
+	mux = http.NewServeMux()
+
+	apiHandler := http.NewServeMux()
+	apiHandler.Handle(baseUrl+"/", http.StripPrefix(baseUrl, mux))
+
+	// server is a test HTTP server used to provide mock API responses.
+	server := httptest.NewServer(apiHandler)
+
+	// client is the GitHub client being tested and is
```
Comment (chivalryq) [inline pkg/addon/reader_gitlab_test.go:47]: type

**Review by chivalryq — COMMENTED** [review]
S

**Comment by barnettZQG** [comment]
@lgj101 can you contribute the UI to support the GitLab type? Ref: https://github.com/oam-dev/velaux/blob/main/src/pages/Addons/components/registry-manage/index.tsx

**Comment by chivalryq** [comment]
LGTM after fix above. Nice job!

**Review by fourierr — COMMENTED** [review]
LGTM in general,

File: pkg/utils/parse.go [hunk 7]
```diff
@@ -198,3 +212,24 @@ func ByteCountIEC(b int64) string {
...
+	}
+	repo := strings.Split(addr, owner)
+
+	// https://example.gitlab.com/<owner>/<repo>
+	return TypeGitlab, &Content{
+		GitlabContent: GitlabContent{
+			Host:  URL.Scheme + "://" + URL.Host,
+			Owner: owner,
+			Repo:  repo[1][1:],
+			Ref:   "", // use default branch
```
[the remark this answers is not in the document]
Comment (lgj101) [reply pkg/utils/parse.go:232]: 可以的，后续我会加上

**Comment by lgj101** [comment]
> @lgj101 can you contribute the UI to support the GitLab type? Ref: https://github.com/oam-dev/velaux/blob/main/src/pages/Addons/components/registry-manage/index.tsx

好的  我后面加过去

**Comment by lgj101** [comment]
> @lgj101 can you contribute the UI to support the GitLab type? Ref: https://github.com/oam-dev/velaux/blob/main/src/pages/Addons/components/registry-manage/index.tsx

**Review by chivalryq — APPROVED** [review]

**Comment by lgj101** [comment]
> @lgj101 can you contribute the UI to support the GitLab type? Ref: https://github.com/oam-dev/velaux/blob/main/src/pages/Addons/components/registry-manage/index.tsx

https://github.com/oam-dev/velaux/pull/423   已经提了

**Comment by lgj101** [comment]
@chivalryq  Can it be passed?

**Review by fourierr — APPROVED** [review]

**Comment by chivalryq** [comment]
> @chivalryq Can it be passed? 

I have approved this PR yesterday :)

**Comment by lgj101** [comment]
> 
OK

**Review by wonderflow — APPROVED** [review]

File: pkg/utils/parse.go [hunk 8]
```diff
@@ -198,3 +212,40 @@ func ByteCountIEC(b int64) string {
...
+	// We support two valid format:
+	// 1. https://example.gitlab.com/<owner>/<repo>
+	// 2. https://example.gitlab.com/<owner>/<repo>/tree/<branch>
+	URL, err := url.Parse(addr)
+	if err != nil {
+		return "", nil, err
+	}
+
+	arr := strings.Split(addr, repo)
+	owner := strings.Split(arr[0], URL.Host+"/")
```
Comment (FogDong) [inline pkg/utils/parse.go:227]: It's better to make sure that the length of arr is greater than 1 here.

Comment (lgj101) [reply pkg/utils/parse.go:227]: As long as split succeeds, the arr will be greater than 1

File: pkg/addon/source.go [hunk 9]
```diff
@@ -182,23 +196,63 @@ func NewAsyncReader(baseURL, bucket, subPath, token string, rdType ReaderType) (
...
 		g := r.Git
-		return NewAsyncReader(g.URL, "", g.Path, g.Token, gitType)
+		return NewAsyncReader(g.URL, "", "", g.Path, g.Token, gitType)
 	}
 	if r.Gitee != nil {
 		g := r.Gitee
-		return NewAsyncReader(g.URL, "", g.Path, g.Token, giteeType)
+		return NewAsyncReader(g.URL, "", "", g.Path, g.Token, giteeType)
+	}
+	if r.Gitlab != nil {
```
Comment (wangyikewxgm) [inline pkg/addon/source.go:253]: "switch case" is better.

**Review by wangyikewxgm — APPROVED** [review]
LGTM after fix small the issue.

File: pkg/addon/source.go (hunk 9 above)
Comment (lgj101) [reply pkg/addon/source.go:253]: This place is (R * Registry). What needs to be determined here is whether multiple variables of R are nil. It is not supported to determine the type. If switch is used, an additional type needs to be added to R for determination

Comment (wangyikewxgm) [reply pkg/addon/source.go:253]: switch {
case1 : 
    code here
case2:
   code here
case3:
    code here
}

Comment (lgj101) [reply pkg/addon/source.go:253]: OK

Comment (lgj101) [reply pkg/addon/source.go:253]: > switch { case1 : code here case2: code here case3: code here }
I know the format of switch, but currently (R * Registry) cannot be written with switch. For details, you can see the code in this section

**Review by Somefive — APPROVED** [review]

**Review by yangsoon — APPROVED** [review]

**Pull request merged**